src/coreComponents/functions/MultivariableTableFunctionKernels.hpp

```diff
@@ -0,0 +1,326 @@
+    real64 axisLows[numDims];
+    real64 axisMults[numDims];
+
+    for( int i = 0; i < numDims; ++i )
+    {
+      integer const axisIndex = getAxisIntervalIndexLowMult( coordinates[i],
+                                                             m_axisMinimums[i], m_axisMaximums[i],
+                                                             m_axisSteps[i], m_axisStepInvs[i], m_axisPoints[i],
+                                                             axisLows[i], axisMults[i] );
+      hypercubeIndex += axisIndex * m_axisHypercubeMults[i];
+    }
+
+    interpolatePointWithDerivatives( coordinates,
+                                     getHypercubeData( hypercubeIndex ),
+                                     &axisLows[0], &axisMults[0],
+                                     &m_axisStepInvs[0],
+                                     values,
+                                     derivatives );
+
+  }
+
+protected:
+
+  /**
+   * @brief Get pointer to hypercube data
+   *
+   * @param[in] hypercubeIndex
+   * @return pointer to hypercube data
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  real64 const *
+  getHypercubeData( globalIndex const hypercubeIndex ) const
+  {
+    return &m_hypercubeData[hypercubeIndex * numVerts * numOps];
+  }
+
+  /**
+   * @brief Get the interval index, low and mult values for a given axis coordinate
+   *
+   * @param[in] axisCoordinate coordinate on a given axis
+   * @param[in] axisMin minimum value on a given axis
+   * @param[in] axisMax maximum value on a given axis
+   * @param[in] axisStep interval length for a given axis
+   * @param[in] axisStepInv inversion of the interval length for a given axis
+   * @param[in] axisPoints number of discretization points for a given axis
+   * @param[out] axisLow left coordinate of target axis interval
+   * @param[out] axisMult weight of the right coordinate of target axis interval
+   * @return integer target axis interval index
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  integer
+  getAxisIntervalIndexLowMult( real64 const axisCoordinate,
+                               real64 const axisMin,
+                               real64 const axisMax,
+                               real64 const axisStep,
+                               real64 const axisStepInv,
+                               integer const axisPoints,
+                               real64 & axisLow,
+                               real64 & axisMult ) const
+  {
+    integer axisIntervalIndex = integer((axisCoordinate - axisMin) * axisStepInv );
+
+    // check that axisindex is within interpolation interval: valid axisindex is between 0 and (axisPoints - 2),
+    // since there are axisPoints-1 intervals along the axis
+
+    if( axisIntervalIndex < 0 )
+    {
+      axisIntervalIndex = 0;
+      if( axisCoordinate < axisMin )
+      {
+        printf( "Interpolation warning: axis is out of limits (%lf; %lf) with value %lf, extrapolation is applied\n", axisMin, axisMax, axisCoordinate );
+      }
```

Comment:
Is this going to lead to an overwhelming number of warnings in your experience? Is it worth just stopping the simulation if that happens?

Reply:
No, it will not, I see these messages pretty rare. In most cases, it merely indicates that parameter space limits were chosen inappropriately: for example, imagine you have an injector at 100 bar and producer at 80 bar, so you discretize your pressure from 75 to 105 (just to be sure), and all goes fine. Later, you change your injector to 110, forgetting to change pressure discretization limits - and then you can get the message above. In other cases, if a linear solver (maybe with low tolerance settings) gives a very rough estimation, your nonlinear update can consequently also be inaccurate and throw you out of the discretized space. I don't think it's worth stopping simulation though, because (especially in the second case) it can recover on its own. I am only changing output syntax to more appropriate rather than pure printf.